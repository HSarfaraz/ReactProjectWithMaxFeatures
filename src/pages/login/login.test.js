import { render, screen } from "@testing-library/react";
// import { fireEvent, render, screen } from '@testing-library/react'
import userEvent from "@testing-library/user-event";
import { BrowserRouter as Router } from "react-router-dom";

import Login from "./login";
import { createMemoryHistory } from "history";

describe("Test the login component", () => {
  //test for directing route const navigate = useNavigate();

  test("should redirect and update history", () => {
    // dont check with history
    const history = createMemoryHistory();
    render(
      <Router>
        <Login />
      </Router>
    );

    const userClick = screen.getByRole("button", { name: "Log in" });
    userEvent.click(userClick);

    expect(history.location.pathname).toEqual("/"); //make it fail
  });

  it("should render login component correctly", async () => {
    const { asFragment } = render(
      <Router>
        <Login />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  function renderWithRouter(
    ui,
    {
      route = "/",
      history = createMemoryHistory({ initialEntries: [route] }),
    } = {}
  ) {
    return {
      ...render(<Router history={history}>{ui}</Router>),
      history,
    };
  }

  it("should allow to navigate to dashboard if login button is clicked", () => {
    const { history } = renderWithRouter(<Login />, {
      route: "/login",
    });
    // expect(history.location.pathname).toBe('/login');
    const userClick = screen.getByRole("button", { name: "Log in" });
    userEvent.click(userClick);
    expect(history.location.pathname).toBe("/login");
  });

  it("login component should not be an empty element", () => {
    const { container } = render(
      <Router>
        <Login />
      </Router>
    );
    expect(container).not.toBeEmptyDOMElement();
  });

  test("Show error message if email or user id is wrong", () => {
    render(
      <Router>
        <Login />
      </Router>
    );

    const emailNode = screen.getByPlaceholderText("Enter Email");
    userEvent.type(emailNode, "michael.lawson@gmail.com");

    const userClick = screen.getByRole("button", { name: "Log in" });
    userEvent.click(userClick);

    const message = screen.getByText(/Please enter mycompany id/); //add the data-test-id

    expect(message).toBeInTheDocument();
  });

  //write after validation, if wrong user id, show the error
  test("email input feild should accept email", () => {
    const onLogin = jest.fn();
    render(
      <Router>
        <Login onLogin={onLogin} />
      </Router>
    );
    const emailNode = screen.getByPlaceholderText("Enter Email");
    userEvent.type(emailNode, "michael.lawson@mycompany.com");
    expect(emailNode.value).toMatch("michael.lawson@mycompany.com");
  });

  // test based on backend functionality
});
