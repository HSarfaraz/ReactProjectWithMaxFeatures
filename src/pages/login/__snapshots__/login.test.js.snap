// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the login component should render login component correctly 1`] = `
<DocumentFragment>
  <div>
    <form>
      <div
        className="sc-bczRLJ sc-dkzDqf bqrCNG ejwePW"
      >
        <header
          className="sc-hKMtZM iGbshS"
        >
          <h1
            className="sc-jSMfEi emWAPi"
          >
            Log in
          </h1>
        </header>
        <div
          className="sc-gKXOVf dFeWCX"
        >
          <fieldset
            className="sc-iBkjds cXEpGZ"
          >
            <label>
              Email address
            </label>
            <input
              className="sc-ftvSup eicDkB"
              placeholder="Enter Email"
              type="email"
              value=""
            />
          </fieldset>
          <fieldset
            className="sc-iBkjds cXEpGZ"
          >
            <label>
              Password
            </label>
            <input
              className="sc-ftvSup eicDkB"
              placeholder="Enter Password"
              type="password"
              value=""
            />
            <span
              className="sc-papXJ jZhlPj fa fa-eye"
            />
          </fieldset>
          <fieldset
            className="sc-iBkjds cXEpGZ"
          >
            <button
              className="sc-crXcEl gGqYCz"
              type="submit"
            >
              Log in
            </button>
          </fieldset>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
